### PR TITLE
drivers: dai: intel: ssp: RX FIFO drain rework and DMA request management changes (take 2)

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

#62596 caused regression in internal CI test and got reverted, however the RX FIFO rework is needed for error free operation.
The root cause of the failure has been identified and fixed by a new patch (the first one in the PR):
The generated blob that is used by the internal CI had the RX/TX enable bits set along with the RX/TX DMA request enable bits. This is not correct and it will be fixed in the blob but the driver should be able to handle such cases by ignoring these enable bits from the blob.
The SSP driver manages both the RX/TX enable and DMA request bits itself.

The other new patch in the series is the change on how the DMA request is managed. There is no reason to enable both at the same time since if we have only TX then why would we have the RX DMA request enabled?
The DMA request has nothing to do with clock generation either.

This also makes the code a bit easier to follow because we don't have the DMA request toggling scattered around.

The PR is passing now both SOF CI and Internal CI (5 test runs has been done to make sure that we will not end up reverting again).

### CODE REVIEW

**Review by kv2019i — DISMISSED** [review]
Thanks @ujfalusi , looks good now. You do need to fix the compliance check issue with 100+ line, this has to pass. Rest looks good.

**Comment by ujfalusi** [comment]
changes since v1:
- break long line to fix compliance check.

**Comment by ujfalusi** [comment]
> Thanks @ujfalusi , looks good now. You do need to fix the compliance check issue with 100+ line, this has to pass. Rest looks good.

Done and the 6th internal and SOF CI check is also passed.

**Comment by ujfalusi** [comment]
@kv2019i, the compliance issue is fixed now.

**Review by kv2019i — APPROVED** [review]

**Review by lyakh — APPROVED** [review]

File: drivers/dai/intel/ssp/ssp.c [hunk 1]
```diff
@@ -1759,15 +1785,25 @@ static void dai_ssp_set_reg_config(struct dai_intel_ssp *dp, const struct dai_co
...
 	ssc0 = regs->ssc0;
-	sstsa = regs->sstsa;
-	ssrsa = regs->ssrsa;
+	sstsa = SSTSA_GET(regs->sstsa);
+	ssrsa = SSRSA_GET(regs->ssrsa);
+	sscr1 = regs->ssc1 & ~(SSCR1_RSRE | SSCR1_TSRE);
+
+	if (regs->sstsa & SSTSA_TXEN || regs->ssrsa & SSRSA_RXEN ||
+	    regs->ssc1 & (SSCR1_RSRE | SSCR1_TSRE)) {
+		LOG_INF("%s: Ignoring %s%s%s%sfrom blob", __func__,
```
Comment (marc-hb) [inline drivers/dai/intel/ssp/ssp.c:1797]: @ujfalusi we're seeing this error message pretty much everywhere:
- https://github.com/thesofproject/sof/issues/9026

Any idea why?

Comment (marc-hb) [reply drivers/dai/intel/ssp/ssp.c:1797]: Sorry the LOG_ERR was in commit 6423bc3bc885 actually.

Comment (ujfalusi) [reply drivers/dai/intel/ssp/ssp.c:1797]: @marc-hb, no, this print remained LOG_INF:
https://github.com/zephyrproject-rtos/zephyr/commit/6423bc3bc8853997e6974cf2aa68c4c4ee2c3abe#diff-69c3bde06e9c145ab8321fcce947ac6d765aa2ffe873641125e6d85d41f45943L1797-R1770

The blob is incorrect, it has the TX/RX enabled and/or the TX/RX DMA enabled. This is invalid and it is ignored, we don't want to enable the SSP when we load the blob, we want to enable it when it needs to be enabled.
We can lower the level, but the blob should be fixed.

Comment (marc-hb) [reply drivers/dai/intel/ssp/ssp.c:1797]: > no, this print remained LOG_INF:

Yes that's what I meant, sorry for the confusion.

> We can lower the level, but the blob should be fixed.

So should the "Ignoring ..." line be a WARN maybe?

Comment (ujfalusi) [reply drivers/dai/intel/ssp/ssp.c:1797]: I would leave it to INF, the blob might be coming from BIOS/ROM and not from tplg, it is hard to fix those.

**Pull request merged**